File: R/ChrDivi.R
ChrDivi <- R6::R6Class(
  classname = "ChrDivi",
  inherit = BaseProc,
  public = list(
    initialize = function(atacProc, ReadsIfile = NULL, ReadsOpath = NULL, name = NULL,
                          sort = NULL, sort_col = NULL, editable = FALSE){
      super$initialize("ChrDivi", editable, list(arg1 = atacProc))

      # necessary parameters
      private$paramlist[["ReadsIfile"]] <- ReadsIfile
      private$paramlist[["sort"]] <- sort
      private$paramlist[["sort_col"]] <- sort_col
      # unnecessary parameters
      if(is.null(ReadsOpath)){
        private$paramlist[["ReadsOpath"]] <- paste0(dirname(private$paramlist[["ReadsIfile"]]),
                                                    "/", collapse = "")
      }else{
        private$paramlist[["ReadsOpath"]] <- ReadsOpath
        # add "/" in the output path
        if(substr(private$paramlist[["ReadsOpath"]], nchar(private$paramlist[["ReadsOpath"]]),
                  nchar(private$paramlist[["ReadsOpath"]])) != "/"){
          private$paramlist[["ReadsOpath"]] <- paste0(private$paramlist[["ReadsOpath"]],
                                                      "/", collapse = "")
        }
      }
      if(is.null(name)){
        private$paramlist[["name"]] <- "Output"
      }else{
        private$paramlist[["name"]] <- name
      }

      # parameter check
      private$paramValidation()
    } # initialization end

  ), #public end

  private = list(
    processing = function(){
      private$writeLog(paste0("processing file:"))
      private$writeLog(sprintf("source:%s", private$paramlist[["ReadsIfile"]]))
      private$writeLog(sprintf("destination:%s", private$paramlist[["ReadsOpath"]]))
      .chr_separate_call(ReadsIfile = private$paramlist[["ReadsIfile"]], ReadsOpath = private$paramlist[["ReadsOpath"]],
                         Name = private$paramlist[["name"]])
      if(!is.null(private$paramlist[["sort"]]) && (private$paramlist[["sort"]])){
        chr <- as.list(c(1:22, "X", "Y"))
        for(i in seq(1:24)){
          sfile <- paste(private$paramlist[["ReadsOpath"]], private$paramlist[["name"]],
                         "_chr", chr[[i]], ".bed", sep = "", collapse = "")
          data <- read.table(sfile)
          data <- data[order(data[, private$paramlist[["sort_col"]]]), ]
          write.table(data, file = sfile, row.names = FALSE, col.names = FALSE, quote = FALSE)
        }
      }
    }, # processing end

    checkRequireParam = function(){
      if(is.null(private$paramlist[["ReadsIfile"]])){
        stop("Parameter ReadsIfile is required!")
      }
      if(!is.null(private$paramlist[["sort"]]) && (private$paramlist[["sort"]])){
        if(is.null(private$paramlist[["sort_col"]])){
          stop("Please specify which column you want sort!")
        }
      }
    }, # checkRequireParam end

    checkAllPath = function(){
      private$checkFileExist(private$paramlist[["ReadsIfile"]])
      private$checkPathExist(private$paramlist[["ReadsOpath"]])
    } # checkAllPath end

  ) # private end

) # class end


#' separate genome information by chromatin name.
#' @param atacProc Not using now, we will use it in the future.
#' @param Ifile Input bed file path, the first column is chromatin name, must be sorted by chromatin.
#' @param Opath The output path, an empty folder would be great.
#' If not specified, the output file will be writen in ReadsIfile's dirctionary.
#' @param prefix the prefix of the output name, format:prefix_chr*.bed, default:output.
#' @param sort TRUE or FALSE, sort every output file by a column.
#' @param sort_col Which column you want to sort, if sort = TRUE, this parameter must be specified.
#' @export
atacChrDivi <- function(atacProc = NULL, Ifile = NULL, Opath = NULL,
                        prefix = NULL, sort = NULL, sort_col = null){
  tmp <- ChrDivi$new(atacProc, Ifile, Opath, prefix, sort, sort_col)
  tmp$process()
  return(tmp)
}

File: src/ChrDivi.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <cstring>
#include <sstream>
#include <vector>
#include <stdlib.h>
#include "ChrDivi.h"
#include "RcoutRcerr.h"
using namespace std;

ChrInfoDivi::ChrInfoDivi(string readsIfile, string readsOpath, string Outputname)
{
  this -> readsIfile = readsIfile;
  this -> readsOpath = readsOpath;
  this -> Outputname = Outputname;
}

int ChrInfoDivi::DoDivi()
{
  // input bed file
  string ipath = this -> readsIfile;
  // output path + prefix
  string opath = this -> readsOpath;
  string name = this -> Outputname;
  ifstream readsifile(ipath.c_str(), ios::in);

  // vector to save string
  vector<string> content;

  // parameters used in this program
  char line[100000] = {0};
  const char *sep = "\t ";

  // initialization
  if(!readsifile.getline(line, sizeof(line)))
  {
    cout<<"ERROR: the input file is empty!"<<endl;
    return 0;
  }
  string tmp_line = line;
  string chr(strtok(line, sep));
  string chr_flag;
  chr_flag = chr;
  string outputfile = opath + name + "_" + chr + ".bed";
  content.push_back(tmp_line);

  while(readsifile.getline(line, sizeof(line)))
  {
    tmp_line = line;
    chr = strtok(line, sep);
    if(chr == chr_flag)
    {
      content.push_back(tmp_line);
    }
    else
    {
      //********************write data to the output file********************
      ofstream readsofile(outputfile.c_str(), ios::out);
      int ctsize = content.size();
      for(int i = 0; i < ctsize; i++)
      {
        readsofile << content[i] << "\n";
      }
      readsofile.close();
      content.clear();
      //*********************************************************************
      chr_flag = chr;
      outputfile = opath + name + "_" + chr + ".bed";
      content.push_back(tmp_line);
    }
  }

  //********************write the last chromatin data to the output file********************
  ofstream readsofile(outputfile.c_str(), ios::out);
  int ctsize = content.size();
  for(int i = 0; i < ctsize; i++)
  {
    readsofile << content[i] << "\n";
  }
  readsofile.close();
  //****************************************************************************************

  readsifile.close();
  return 0;
}
